File: python/heritage/domain/portfolio_aggregate.py
# ============================================================
# 文件：portfolio_aggregate.py
# 职责：维护单账户的完整投资组合状态，并计算风险指标快照。
#
# 领域聚合根（Domain Aggregate Root）：
#   封装账户内所有持仓、盈亏状态，
#   保证内部数据一致性（持仓变更必须通过 apply_trade() 进行）。
#
# 持仓更新策略：加权平均成本法（Average Cost Basis）
#   - 增仓：加权平均更新成本
#   - 减仓：按平均成本计算已实现盈亏
# ============================================================

from __future__ import annotations
import math
from events.events import TradeExecutedEvent, Side
from .instrument import Option
from .pricing_engine import IPricingEngine
from .risk_metrics import RiskMetrics


class PortfolioAggregate:
    """单账户投资组合聚合根。"""

    def __init__(self, account_id: str, options: list[Option]) -> None:
        self._account_id      = account_id        # 账户标识符
        self._options         = options           # 期权合约列表（用于 Greeks 计算）
        self._positions:  dict[str, int]   = {}   # 净持仓
        self._avg_cost:   dict[str, float] = {}   # 平均成本
        self._realized_pnl    = 0.0               # 累计已实现盈亏
        self._unrealized_pnl  = 0.0               # 当前未实现盈亏
        self._total_pnl_high  = 0.0               # 日内总盈亏高水位

    def apply_trade(self, event: TradeExecutedEvent) -> None:
        """应用成交事件，更新净持仓与已实现盈亏。

        客户视角转换：
          客户 Buy  → 做市商持仓减少（做市商卖出）
          客户 Sell → 做市商持仓增加（做市商买入）
        """
        id_      = event.instrument_id
        # 做市商持仓变化方向
        pos_chg  = +event.quantity if event.side == Side.Sell else -event.quantity
        old_pos  = self._positions.get(id_, 0)
        new_pos  = old_pos + pos_chg
        price    = event.price

        if old_pos == 0 or (old_pos > 0 and pos_chg > 0) or (old_pos < 0 and pos_chg < 0):
            # 同向加仓：加权平均更新成本
            old_notional = self._avg_cost.get(id_, price) * abs(old_pos)
            new_notional = price * abs(pos_chg)
            total_qty    = abs(old_pos) + abs(pos_chg)
            self._avg_cost[id_] = (old_notional + new_notional) / total_qty if total_qty else price
        else:
            # 反向减仓：计算已实现盈亏
            close_qty = min(abs(old_pos), abs(pos_chg))
            cost      = self._avg_cost.get(id_, price)
            pnl = (price - cost) * close_qty if old_pos > 0 else (cost - price) * close_qty
            self._realized_pnl += pnl
            if abs(pos_chg) > abs(old_pos):
                # 反手建仓，更新成本为新价格
                self._avg_cost[id_] = price

        self._positions[id_] = new_pos
        print(
            f"[持仓聚合|{self._account_id}] {id_} 持仓更新: "
            f"{old_pos:+d} → {new_pos:+d}  已实现盈亏累计: ${self._realized_pnl:.2f}"
        )

    def mark_to_market(self, engine: IPricingEngine, underlying_price: float) -> None:
        """按当前市价重新估值，更新未实现盈亏与高水位。"""
        unrealized = 0.0
        for opt in self._options:
            pos = self.get_position(opt.id)
            if pos == 0:
                continue
            result = engine.price(opt, underlying_price)
            cost   = self._avg_cost.get(opt.id, result.theo)
            # 未实现盈亏 = (当前价 - 成本价) × 净持仓
            unrealized += (result.theo - cost) * float(pos)

        self._unrealized_pnl = unrealized
        # 更新日内总盈亏高水位（用于回撤计算）
        total_pnl            = self._realized_pnl + self._unrealized_pnl
        self._total_pnl_high = max(self._total_pnl_high, total_pnl)

    def compute_metrics(
        self, engine: IPricingEngine, underlying_price: float
    ) -> RiskMetrics:
        """生成当前时刻的风险指标快照。"""
        m = RiskMetrics(
            realized_pnl   = self._realized_pnl,
            unrealized_pnl = self._unrealized_pnl,
        )

        # 计算组合 Delta：Σ(delta_i × position_i)
        for opt in self._options:
            pos = self.get_position(opt.id)
            if pos == 0:
                continue
            result  = engine.price(opt, underlying_price)
            m.delta += result.delta * float(pos)

        # 日内最大回撤 = 高水位 - 当前总盈亏
        total_pnl          = self._realized_pnl + self._unrealized_pnl
        m.intraday_drawdown = self._total_pnl_high - total_pnl

        # VaR 简化估算：1日 VaR ≈ |Delta| × S × σ × √(1/252)（95% 置信度，σ=20% 代理值）
        m.var_1d = abs(m.delta) * underlying_price * 0.20 / math.sqrt(252.0)

        return m

    def get_position(self, instrument_id: str) -> int:
        """查询指定合约净持仓。"""
        return self._positions.get(instrument_id, 0)

File: src/core/domain/PortfolioAggregate.cpp
#include "PortfolioAggregate.hpp"
#include <iostream>
#include <iomanip>
#include <algorithm>
#include <cmath>

// ============================================================
// PortfolioAggregate 实现
//
// 设计说明：
//   持仓更新采用"净持仓 + 平均成本"模型（Average Cost Basis）：
//   - 增仓时：加权平均更新平均成本
//   - 减仓时：按平均成本计算已实现盈亏，不改变剩余仓位的成本
// ============================================================

namespace omm::domain {

PortfolioAggregate::PortfolioAggregate(
    std::string account_id,
    std::vector<std::shared_ptr<Option>> options)
    : account_id_(std::move(account_id))
    , options_(std::move(options)) {}

void PortfolioAggregate::applyFill(const events::FillEvent& event) {
    const std::string& id = event.instrument_id;

    // 我方视角（Phase 2 统一语义）：
    //   Buy  → 我们买入 → 持仓增加
    //   Sell → 我们卖出 → 持仓减少
    int pos_change = (event.side == events::Side::Buy)
                     ? +event.fill_qty   // 我们买入：多头增加
                     : -event.fill_qty;  // 我们卖出：空头增加

    int old_pos  = positions_[id];
    int new_pos  = old_pos + pos_change;
    double price = event.fill_price;

    if (old_pos == 0 || (old_pos > 0 && pos_change > 0) || (old_pos < 0 && pos_change < 0)) {
        // 同向加仓：加权平均更新成本
        double old_notional = avg_cost_[id] * std::abs(old_pos);
        double new_notional = price * std::abs(pos_change);
        int    total_qty    = std::abs(old_pos) + std::abs(pos_change);
        avg_cost_[id] = (total_qty > 0)
                        ? (old_notional + new_notional) / total_qty
                        : price;
    } else {
        // 反向减仓：计算已实现盈亏
        int close_qty = std::min(std::abs(old_pos), std::abs(pos_change));
        // 做市商持有多头时：平仓盈亏 = (平仓价 - 成本价) × 数量
        // 做市商持有空头时：平仓盈亏 = (成本价 - 平仓价) × 数量
        double pnl = (old_pos > 0)
                     ? (price - avg_cost_[id]) * close_qty
                     : (avg_cost_[id] - price) * close_qty;
        realized_pnl_ += pnl;

        if (std::abs(pos_change) > std::abs(old_pos)) {
            // 反手：先平仓再反向建仓，更新成本为新价格
            avg_cost_[id] = price;
        }
        // 若完全平仓，保留 avg_cost_（下次建仓时会覆盖）
    }

    positions_[id] = new_pos;

    std::cout << "[持仓聚合|" << account_id_ << "] "
              << id << " 持仓更新: "
              << std::showpos << old_pos << " → " << new_pos
              << std::noshowpos
              << "  已实现盈亏累计: $"
              << std::fixed << std::setprecision(2) << realized_pnl_ << "\n";
}

void PortfolioAggregate::markToMarket(
    const IPricingEngine& engine, double underlying_price) {

    double unrealized = 0.0;

    for (const auto& opt : options_) {
        int pos = get_position(opt->id());
        if (pos == 0) continue;

        // 按 Black-Scholes / 简化模型获取当前理论价格
        PriceResult result = engine.price(*opt, underlying_price);
        double cost        = avg_cost_[opt->id()];

        // 未实现盈亏 = (当前价 - 成本价) × 净持仓
        // 注意：做市商多头 pos>0 时希望价格上涨；空头 pos<0 时希望价格下跌
        unrealized += (result.theo - cost) * static_cast<double>(pos);
    }

    unrealized_pnl_ = unrealized;

    // 更新日内总盈亏高水位（用于回撤计算）
    double total_pnl = realized_pnl_ + unrealized_pnl_;
    total_pnl_high_  = std::max(total_pnl_high_, total_pnl);
}

RiskMetrics PortfolioAggregate::computeMetrics(
    const IPricingEngine& engine, double underlying_price) const {

    RiskMetrics m;
    m.realized_pnl   = realized_pnl_;
    m.unrealized_pnl = unrealized_pnl_;

    // 计算组合 Delta：Σ(delta_i × position_i)
    for (const auto& opt : options_) {
        int pos = get_position(opt->id());
        if (pos == 0) continue;
        PriceResult result = engine.price(*opt, underlying_price);
        m.delta += result.delta * static_cast<double>(pos);
    }

    // 日内最大回撤 = 高水位 - 当前总盈亏
    double total_pnl      = realized_pnl_ + unrealized_pnl_;
    m.intraday_drawdown   = total_pnl_high_ - total_pnl;

    // Gamma、Vega、Theta 此处简化为 0（SimplePricingEngine 不提供这些值）
    // 接入完整 BS 引擎后可在此扩展
    m.gamma = 0.0;
    m.vega  = 0.0;
    m.theta = 0.0;

    // VaR 简化估算：1日 VaR ≈ |Delta| × S × σ × √(1/252) （95% 置信度）
    // 此处使用固定波动率 20% 作为占位符
    const double vol_proxy = 0.20;
    m.var_1d = std::abs(m.delta) * underlying_price * vol_proxy
               / std::sqrt(252.0);

    return m;
}

int PortfolioAggregate::get_position(const std::string& instrument_id) const {
    auto it = positions_.find(instrument_id);
    return (it != positions_.end()) ? it->second : 0;
}

} // namespace omm::domain
